lib/Onnxifi/GlowOnnxManager.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "GlowOnnxManager.h"
+
+#include "Base.h"
+
+#include <mutex>
+
+namespace glow {
+namespace onnxifi {
+GlowOnnxManager &GlowOnnxManager::get() {
+  static GlowOnnxManager manager;
+  return manager;
+}
+
+void GlowOnnxManager::addBackendId(BackendIdPtr backendId) {
+  assert(!isValid(backendId));
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    backendIds_.insert(backendId);
+  }
+  assert(isValid(backendId));
+}
+
+BackendPtr GlowOnnxManager::createBackend(BackendIdPtr backendId) {
+  assert(isValid(backendId));
+  BackendPtr backend = new Backend(backendId);
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    backends_.insert(backend);
+  }
+  assert(isValid(backend));
+  return backend;
+}
+
+EventPtr GlowOnnxManager::createEvent() {
+  EventPtr event = new glow::onnxifi::Event();
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    events_.insert(event);
+  }
+  assert(isValid(event));
+  return event;
+}
+
+GraphPtr GlowOnnxManager::createGraph(BackendPtr backend) {
+  assert(isValid(backend));
+  GraphPtr graph = new glow::onnxifi::Graph(backend);
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    graphs_.insert(graph);
+  }
+  assert(isValid(graph));
+  return graph;
+}
+
+bool GlowOnnxManager::isValid(BackendIdPtr backendId) {
+  std::lock_guard<std::mutex> lock(m_);
```

Comment:
This won't work. You will have a recursive lock?

Reply:
@yinghai why won't this work?

Reply:
OK. This works, as you don't have recursive locks. :) But this kind of things is a bit error prone...

Reply:
Imo with `lock_guard` it's reasonably straightforward to see the scope of the locking and avoid errors. What would you view as being less prone to errors here?

Reply:
When you have a function that grabs a lock and calls a function that grabs the same lock, it's kind of tricky to to explicitly manage the lock_guard to avoid recursive lock, like what you did here. It's also like three lock/unlock for a function. Ideally, one would be nice. BTW, in opt mode, none of the checks will generate actual code and we essentially will have only one lock, right?